core/src/main/scala/kafka/security/authorizer/AclAuthorizer.scala

```diff
@@ -188,7 +188,9 @@ class AclAuthorizer extends Authorizer with Logging {
     val deleteExceptions = new mutable.HashMap[AclBinding, ApiException]()
     val filters = aclBindingFilters.asScala.zipWithIndex
     inWriteLock(lock) {
-      val resourcesToUpdate = aclCache.keys.map { resource =>
+      // Find all potentially matching resource patterns from the provided filters and ACL cache and apply the filters
+      val resources = aclCache.keys ++ filters.map(_._1.patternFilter).filter(_.matchesAtMostOne).flatMap(filterToResources)
+      val resourcesToUpdate = resources.map { resource =>
```

Comment:
Another thing to consider for a separate PR: this method seems highly inefficient. We take _all_ the keys in the cache and do a bunch of transformations. It seems like we should avoid that, no? Intuitively, we would do a filter operation first and then transform only the cache items that match.

Reply:
I will do https://issues.apache.org/jira/browse/KAFKA-8847 first to remove references to the old classes and deprecate those. And that would avoid a lot of the unnecessary conversions. I can see if more can be done to improve this in the same PR.

Reply:
Sounds good.